repair/repair.cc

```diff
@@ -997,8 +1012,22 @@ future<> repair::shard_repair_task_impl::run() {
     auto remove_shard_task_id = defer([this] {
         rs.get_repair_module().remove_shard_task_id(global_repair_id.id);
     });
+
     try {
-        co_await do_repair_ranges();
+        // Repair tables in the keyspace one after another
+        assert(table_names().size() == table_ids.size());
+        for (size_t idx = 0; idx < table_ids.size(); idx++) {
+            table_info table_info{
+                .name = table_names()[idx],
+                .id = table_ids[idx],
+            };
+            // repair all the ranges in limited parallelism
+            rlogger.info("repair[{}]: Started to repair {} out of {} tables in keyspace={}, table={}, table_id={}, repair_reason={}",
+                    global_repair_id.uuid(), idx + 1, table_ids.size(), _status.keyspace, table_info.name, table_info.id, _reason);
+
+            auto task = co_await rs.get_repair_module().make_and_start_task<table_repair_task_impl>(get_repair_uniq_id().task_info, *this, table_info, _reason);
+            co_await task->done();
```

Comment:
We pass a reference to table_repair_task_impl and here we wait for the done, so the ref will be valid. But what if a user does not wait for the task done inside the shard task and destroy the shard task before the range task, won't we have use after free?

Reply:
A user cannot unregister the task before it is complete and it won't be complete until all its children are.

There is a use after free danger in get_progress though, I will fix it.